install.c

```diff
@@ -1,29 +1,78 @@
 #include "substdio.h"
 #include "strerr.h"
+#include "env.h"
 #include "error.h"
+#include "fifo.h"
 #include "open.h"
 #include "readwrite.h"
 #include "exit.h"
+#include "alloc.h"
+#include "str.h"
+#include "stralloc.h"
 
 extern void hier();
 
 #define FATAL "install: fatal: "
 
 int fdsourcedir = -1;
 
+static void die_nomem()
+{
+  strerr_die2sys(111,FATAL,"out of memory");
+}
+
+static int ddhome(stralloc *dd, const char *home)
+{
+  const char *denv = env_get("DESTDIR");
+  if (!denv)
+    return 0;
+
+  if (!stralloc_copys(dd,denv)) die_nomem();
+  if (!stralloc_catb(dd,home,str_len(home))) die_nomem();
```

Comment:
It might be worth adding the following above this line:

    if (home[0] != '/') {
      if (!stralloc_copys(dd,"/")) die_nomem();
    }

In the event that the first line in conf-qmail isn't slash (/).

Reply:
This has been broken forever, I don't think this could ever work.

Reply:
I'm not sure what you're referring to, as regards what is broken.  As best as I can tell DESTDIR with or without a trailing slash (/) works here, because home (default: /var/qmail) has a leading slash.  Reading through the mkdir_p implementation, I found it to work if DESTDIR had a trailing slash, which would lead to a double slash (//) in `dd`.  That can leading to one or more trailing slashes in the argument passed to mkdir(2), that syscall tolerates that condition.

If you were referring to something else please excuse the noise.  I did find this behavior subtle but couldn't find anything wrong except in the case when conf-qmail is a relative path *and* DESTDIR is set, which is the genesis of my comment about adding a slash in that case.  In practice I doubt this comes up, much as it's worth fixing.

Reply:
I think that a relative path in conf-qmail would never work, as all programs try to do ```chdir(auto_qmail)``` at the beginning, which would fail for a relative homedir. DESTDIR with a trailing / would not do anything bad, worst is that it would try to create the same directory twice. There was a race condition in that code, which I will remove in a second.